### DISCUSSION TITLE
Dynamically importing component from outside source

### DISCUSSION THREAD

**Discussion by tessamerrill** [vercel/next.js#37440] [Help]
I'm having some slight trouble with dynamically importing a component from an outside server (simple Golang file server).

I'm able to load the file through `next/dynamic` and `import()`, however it gives me the error:

`expected expression, got '<'`

I assume this is some kind of Content-Type issue, however everything I have tried hasn't seemed to work. (current Content-Type is set to `text/javascript`, so that's why I assumed it was that)

Code:
```ts
export default function TemplatePage({ page }: { page: Page }) {
  const [Components, setComponents] = useState<any[]>([]);

  useEffect(() => {
    page.children.map((child) => {
      setComponents((prev) => [
        ...prev,
        dynamic(
          () =>
            import(
              /* webpackIgnore: true */ `http://127.0.0.1:4567/${child.component}.jsx`
            )
        ),
      ]);
    });
  }, []);

  return (
    <div>
      {Components.map((Component, index) => {
        console.log(Component);
        return <Component key={index} />;
      })}
    </div>
  );
}
```

**Comment by icyJoseph** [comment]
This looks interesting, would you mind sharing a repo where this is setup?

I imagine the issue is that the file has `JSX`, and that's not something browsers can really parse. I'd say, transpile that `JSX` file, to `JS` and serve that instead.

Again, I'd love to see a repo with this in action.

**Reply by tessamerrill** [reply]
> This looks interesting, would you mind sharing a repo where this is setup?
> 
> 
> 
> I imagine the issue is that the file has `JSX`, and that's not something browsers can really parse. I'd say, transpile that `JSX` file, to `JS` and serve that instead.
> 
> 
> 
> Again, I'd love to see a repo with this in action.

Yeah I'll see if I can put it in a repo. I asked on the Discord server and effectively came to the conclusion that I'd need to transpile the components.

**Reply by icyJoseph** [reply]
Cool, yeah, I also want to see if this would work, in spite of:

> Note: In import('path/to/component'), the path must be explicitly written. It can't be a template string nor a variable. Furthermore the import() has to be inside the dynamic() call for Next.js to be able to match webpack bundles / module ids to the specific dynamic() call and preload them before rendering. dynamic() can't be used inside of React rendering as it needs to be marked in the top level of the module for preloading to work, similar to React.lazy.

At the end of the day, JS resists everything, so if this works swell, I'd like to add it to my knowledge stash.

**Reply by tessamerrill** [reply]
So I've tried numerous things, multiple libraries, etc. Nothing has seemed to work. I've [made a repository](https://github.com/jackmerrill/sites-renderer) if you want to check it out, and potentially help me haha.

Also the [components server](https://github.com/jackmerrill/sites-components), in case you'll need that.

Current issue I'm having now is `Error: Cannot find module 'http://127.0.0.1:4567/Home.js'`, likely because of the explicit import issue.

So far I've tried:
- Fetching the raw JS/JSX code (both compiled and uncompiled) and parsing that (with both `eval()` and some other libraries)
- Importing a explicit path (as in `http://127.0.0.1:4567/Home.jsx`, no string templating), this is the only one that seemed to work, however it's not what I want.

**Reply by tessamerrill** [reply]
Also have now tried Webpack / Babel, still no luck. Seems the best option is to use just the direct JSX code, as `import()`ing the JSX code compiles it and it works then.

So the issue still remains of the explicit path...

**Reply by tessamerrill** [reply]
(Unless it's possible to import and use a Webpack-compiled component, then maybe)

**Reply by Demolliderish** [reply]
Did you manage to find a solution to this problem?
